PR for 7.6 Beta by making Patchhandler and related classes internal

Sreejithpin opened the pull request:
<!-- markdownlint-disable MD002 MD041 -->

### Issues

*This pull request fixes #xxx.*

### Description

PR for 7.6 Beta by making Patchhandler and related classes internal

Changes: 

DeltaOfT
	internal void Patch(

DeltaSetOfT
	internal DeltaSet<TStructuralType> Patch(IC
	 internal DeltaSet<TStructuralType> Patch(
	 
EdmChangedObjectcollection
	internal EdmChangedObjectCollection Patch(OD

 EdmODataAPIHandler
	internal abstract class EdmODataAPIHandler
	
IODataAPIHandler
	internal interface IODataAPIHandler
	
ODataAPIHandler
	internal abstract class ODataAPIHandler<TStructuralType>: IO
	
ODataAPIHandlerFactory
	internal abstract class ODataAPIHandlerFactory

ODataAPIResponseStatus
	internal enum ODataAPIResponseStatus
	
ODataEdmAPIHandlerFactory	
	internal abstract class ODataEdmAPIHandlerFactory
	
ODataIDResolver
	 internal abstract class ODataIDResolver
	 

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

Sreejithpin committed:
Bulk operations8 (#2502)

* Bulk operations6 (#2)

* BulkInsert1 draft

* Bulk insert changes

* deleted old test

* updates

* Update ODataResourceSetWrapper.cs

* updates

* Updated addressing  comments

* updates

* updates

updates

* Updates

* Updates

* Minor updates

* comments

* updates

* update publicapi for core

* Address comments

* Cleanup and additional tests

Cleanup and additional tests

* Updated code

* BulkInsert1 draft

* Bulk insert changes

* deleted old test

* updates

* updates

* Updated addressing  comments

* updates

* updates

updates

* Updates

* Updates

* Minor updates

* updates

* Address comments

* Cleanup and additional tests

Cleanup and additional tests

* DataAnnotationException updates

* comments

* small updates

* updates

* small update

* updates

* Updates

* Update DeltaSetOfT.cs

* Updates with Patch

* updates

* updates

* Update WebHostTestFixture.cs

* updates

* Update DeltaOfTStructuralType.cs

* Updates

* Updates for serializer etc

* Update WebHostTestFixture.cs

* updates

* updates

* updates

* Bulk Operations Updates

* Review comments addressed

* Updates

* updated for deltaentity

* Update Microsoft.AspNet.OData.Test.csproj

* Update Microsoft.AspNet.OData.PublicApi.bsl

* update public api

* test fix

* Update EdmStructuredObject.cs

* Update EdmStructuredObject.cs

* Update BulkInsertTest.cs

* update

* public api

* review comments

* Review comments updates

* smalll update

* updates

* updates

* updates

Sreejithpin committed:
Bulk operations8 (#2502)

* Bulk operations6 (#2)

* BulkInsert1 draft

* Bulk insert changes

* deleted old test

* updates

* Update ODataResourceSetWrapper.cs

* updates

* Updated addressing  comments

* updates

* updates

updates

* Updates

* Updates

* Minor updates

* comments

* updates

* update publicapi for core

* Address comments

* Cleanup and additional tests

Cleanup and additional tests

* Updated code

* BulkInsert1 draft

* Bulk insert changes

* deleted old test

* updates

* updates

* Updated addressing  comments

* updates

* updates

updates

* Updates

* Updates

* Minor updates

* updates

* Address comments

* Cleanup and additional tests

Cleanup and additional tests

* DataAnnotationException updates

* comments

* small updates

* updates

* small update

* updates

* Updates

* Update DeltaSetOfT.cs

* Updates with Patch

* updates

* updates

* Update WebHostTestFixture.cs

* updates

* Update DeltaOfTStructuralType.cs

* Updates

* Updates for serializer etc

* Update WebHostTestFixture.cs

* updates

* updates

* updates

* Bulk Operations Updates

* Review comments addressed

* Updates

* updated for deltaentity

* Update Microsoft.AspNet.OData.Test.csproj

* Update Microsoft.AspNet.OData.PublicApi.bsl

* update public api

* test fix

* Update EdmStructuredObject.cs

* Update EdmStructuredObject.cs

* Update BulkInsertTest.cs

* update

* public api

* review comments

* Review comments updates

* smalll update

* updates

* updates

* updates

ElizabethOkerio committed:
Feature/new add support for odata bind (#2506)

* support for odata.bind

* support for odata.bind

ElizabethOkerio committed:
Feature/new add support for odata bind (#2506)

* support for odata.bind

* support for odata.bind

Sreejithpin committed:
updates

Sreejithpin committed:
updates

Sreejithpin committed:
small update

Sreejithpin committed:
Code review comments

Sreejithpin committed:
Code review comments

Sreejithpin committed:
update for odataid _temp

Sreejithpin committed:
update for odataid _temp

Sreejithpin committed:
updates

Sreejithpin committed:
updates

Sreejithpin committed:
updates to use only factory

Sreejithpin committed:
updates to use only factory

Sreejithpin committed:
updates

Sreejithpin committed:
updates

Sreejithpin committed:
Squashing Commits

Squashing Commits
Update DeltaOfTStructuralType.cs

test change

Update Microsoft.AspNetCore.OData.Test.csproj

Update Microsoft.Test.E2E.AspNetCore3x.OData.csproj

updates

Update WebStack.versions.settings.targets

Update WebStack.versions.settings.targets

Update GetNugetPackageMetadata.proj

Update WebStack.versions.settings.targets

For testing

update

Update BulkInsertController.cs

Updates

updates

updates

Update NuGet.Config

Update BulkOperationPatchHandlersEF.cs

updates

Sreejithpin committed:
Merged updates

Sreejithpin committed:
Update Microsoft.AspNetCore.OData.PublicApi.bsl

Sreejithpin committed:
Update DeltaTests.cs

Sreejithpin committed:
test update

Sreejithpin committed:
test update

Sreejithpin committed:
test update

Sreejithpin committed:
test update

Sreejithpin committed:
Update DeltaTests.cs

Sreejithpin committed:
Updates for 7.6 release

Sreejithpin committed:
Update WebStack.versions.settings.targets

xuzhg commented on src/Microsoft.AspNet.OData.Shared/DataModificationExceptionType.cs:
```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
```
change all copyrights using existing one.

xuzhg replied:
for other new added files.

xuzhg commented on test/UnitTest/Microsoft.AspNet.OData.Test.Shared/EnableNestedPathsTest.cs:
```diff
@@ -223,7 +223,7 @@ public async Task EnableNestedPaths_Returns404_WhenPathHasUnsupportedSegments(st
             Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
         }
 
-        [Fact]
```
why ship this test case?

Sreejithpin replied:
Test Not relevant anymore as it works differently now and new test cases are added. Removed

xuzhg commented on test/UnitTest/Microsoft.AspNet.OData.Test/app.config:
```diff
@@ -1,7 +1,18 @@
+  <configSections>
+  <!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 -->
+  <section name="entityFramework"
+    type="System.Data.Entity.Internal.ConfigFile.EntityFrameworkSection, EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"
```
why adding these?

xuzhg commented on test/E2ETest/Microsoft.Test.E2E.AspNet.OData/DollarLevels/DollarLevelsEdmModel.cs:
```diff
@@ -18,6 +18,7 @@ public static IEdmModel GetConventionModel(WebRouteConfiguration configuration)
             ODataConventionModelBuilder builder = configuration.CreateConventionModelBuilder();
             builder.EntitySet<DLManager>("DLManagers");
             builder.EntitySet<DLEmployee>("DLEmployees");
+            builder.EntitySet<TestQueryOptions>("Tests");
```
Why do we add new entity set, it's related to Bulkinsert?

xuzhg reviewed (dismissed):
:shipit:

xuzhg commented:
at least rebase and squash merge into one commit.
and need other team member's review ahead.

Sreejithpin commented:
Its already rebased , will sqash and merge wile merging to master

---
In reply to: [985753397](https://github.com/OData/WebApi/pull/2606#issuecomment-985753397) [](http://example.com/codeflow?ancestors=985753397,985753397)

Sreejithpin committed:
updates

Sreejithpin committed:
Update app.config

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// This is the default Patch Handler for non CLR type. This calss has default Get, Create and Update
```
>calss [](http://example.com/codeflow?start=65&length=5)

nit: typo

mikepizzo commented on src/Microsoft.AspNet.OData/GlobalSuppressions.cs:
```diff
@@ -65,4 +65,83 @@
+[assembly: SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "API", Scope = "type", Target = "Microsoft.AspNet.OData.ODataAPIResponseStatus")]
+[assembly: SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "API", Scope = "type", Target = "Microsoft.AspNet.OData.EdmODataAPIHandler")]
+[assembly: SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "API", Scope = "type", Target = "Microsoft.AspNet.OData.ODataAPIHandlerFactory")]
+
```
Are these suppressions still all required?

mikepizzo commented:
> 

Are these suppressions all still required?

---
Refers to: src/Microsoft.AspNet.OData/GlobalSuppressions.cs:148 in 9fd1ffa. [](commit_id = 9fd1ffa1bcc3d42ad3e9252c30e0182d58a7334e, deletion_comment = False)

mikepizzo commented:
>         private EdmDeltaType _edmType;

it appears that this is private field is no longer used, and can be removed?

---
Refers to: src/Microsoft.AspNet.OData.Shared/EdmDeltaDeletedEntityObject.cs:24 in 9fd1ffa. [](commit_id = 9fd1ffa1bcc3d42ad3e9252c30e0182d58a7334e, deletion_comment = False)

mikepizzo commented:
>         private EdmDeltaType _edmType;

it appears that this is private field is no longer used, and can be removed?

---
Refers to: src/Microsoft.AspNet.OData.Shared/EdmDeltaDeletedLink.cs:24 in 9fd1ffa. [](commit_id = 9fd1ffa1bcc3d42ad3e9252c30e0182d58a7334e, deletion_comment = False)

mikepizzo commented:
>         private EdmDeltaType _edmType;

it appears that this is private field is no longer used, and can be removed?

---
Refers to: src/Microsoft.AspNet.OData.Shared/EdmDeltaEntityObject.cs:21 in 9fd1ffa. [](commit_id = 9fd1ffa1bcc3d42ad3e9252c30e0182d58a7334e, deletion_comment = False)

mikepizzo commented:
>         private EdmDeltaType _edmType;

it appears that this is private field is no longer used, and can be removed?

---
Refers to: src/Microsoft.AspNet.OData.Shared/EdmDeltaLink.cs:24 in 9fd1ffa. [](commit_id = 9fd1ffa1bcc3d42ad3e9252c30e0182d58a7334e, deletion_comment = False)

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/IDeltaSet.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    /// <summary>
+    /// Basic interface to reperesent a deltaset which is a collection of Deltas.
+    /// This is being implemented by Deltaset{TStructuralType}. Since its being implementd by a gemeric type and
```
>gemeric [](http://example.com/codeflow?start=96&length=7)

typo

mikepizzo replied:
>Deltaset [](http://example.com/codeflow?start=37&length=8)

Use proper casing -- DeltaSet.
Also, <> instead of {} for generic

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs:
```diff
@@ -358,6 +422,197 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+            try
+            {
+                IODataPathHandler pathHandler = readContext.InternalRequest.PathHandler;
+                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
```
Nit: you can simplify slightly:

                IWebApiRequestMessage internalRequest = readContext.InternalRequest;
                IODataPathHandler pathHandler = internalRequest.PathHandler;

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs:
```diff
@@ -358,6 +422,197 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+
+                string serviceRoot = urlHelper.CreateODataLink(
+                    internalRequest.Context.RouteName,
+                    internalRequest.PathHandler,
```
>internalRequest.PathHandler [](http://example.com/codeflow?start=20&length=27)

just use pathHandler

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,91 @@
+        /// </summary>
+        /// <param name="navigationPath">ODataId in string format</param>
+        /// <param name="pathSegments">Pathsegment collection</param>
+        public NavigationPath(string navigationPath, ReadOnlyCollection<ODataPathSegment> pathSegments)
```
It looks like there is only one case where we set the second parameter, so in all other cases GetNavigationPathItems would return null?

Seems like this class will be hard for the developer to figure out how to use.

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+        {
+            _keys = keys;            
+            _clrType = typeof(TStructuralType);
+            _navigationPath = new NavigationPath(_clrType.Name, null);
```
>_clrType [](http://example.com/codeflow?start=49&length=8)

Is this right?  This will be the type name, not the name of the navigation path, right?

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -107,21 +130,29 @@ public Delta(Type structuralType, IEnumerable<string> updatableProperties)
+            InitializeProperties(updatableProperties);            
+            TransientInstanceAnnotationContainer = new ODataInstanceAnnotationContainer();            
+            _instanceAnnotationsPropertyInfo = instanceAnnotationsPropertyInfo;
+            _navigationPath = new NavigationPath(structuralType.Name, null);
```
>structuralType.Name [](http://example.com/codeflow?start=49&length=19)

is this right? It's the name of the type, not the property

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs:
```diff
@@ -60,6 +74,321 @@ private void Initialize(IEdmEntityType entityType)
             _entityType = entityType;
             _edmType = new EdmDeltaCollectionType(new EdmEntityTypeReference(_entityType, isNullable: true));
             _edmTypeReference = new EdmCollectionTypeReference(_edmType);
+            _navigationPath = new NavigationPath(entityType.Name, null);
```
>entityType.Name [](http://example.com/codeflow?start=49&length=15)

is this right?

gathogojr commented on samples/AspNetODataSample.Web/Web.config:
```diff
@@ -4,85 +4,89 @@
-      <provider invariantName="System.Data.SqlClient" type="System.Data.Entity.SqlServer.SqlProviderServices, EntityFramework.SqlServer" />
-    </providers>
-  </entityFramework>
+	<configSections>
```
Nested increased to 8 spaces. Seems too much to me...

corranrogue9 replied:
Agreed, also I'm not noticing any actual changes to the content of this file.

gathogojr commented on src/Microsoft.AspNet.OData.Shared/Common/SRResources.resx:
```diff
@@ -988,4 +988,22 @@
+    <value>Cannot use Changed Object of type '{0}' on an entity of type '{1}'.</value>
+  </data>
+  <data name="DataModificationException" xml:space="preserve">
+    <value>Core.DataModificationException</value>
```
Do we need a more relevant value here?

Sreejithpin replied:
This is as per design, as per Core Vocab annotations

gathogojr commented on src/Microsoft.AspNet.OData.Shared/Common/SRResources.resx:
```diff
@@ -988,4 +988,22 @@
+    <value>Core.DataModificationException</value>
+  </data>
+  <data name="ContentID" xml:space="preserve">
+    <value>Core.ContentID</value>
```
Do we need a more relevant value here?

Sreejithpin replied:
This is as per design, as per Core Vocab annotations

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DataModificationExceptionType.cs:
```diff
@@ -0,0 +1,119 @@
+        /// <summary>
+        /// Represents response code
+        /// </summary>        
+        public Int16 ResponseCode { get; set; }
```
Why `Int16` and not just `int`?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DataModificationExceptionType.cs:
```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        Unlink
+    }
+}
```
```suggestion
}

```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+            Contract.Assert(entityType != null);
+
+            this.entityType = entityType;
+            this.originalList = originalList?? new List<IEdmStructuredObject>();
```
```suggestion
            this.originalList = originalList ?? new List<IEdmStructuredObject>();
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+        {
+            IEdmNavigationProperty navProperty = entityType.NavigationProperties().FirstOrDefault(navProp => navProp.Name == navigationPropertyName);
+
+            if(navProperty == null)
```
```suggestion
            if (navProperty == null)
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+            IEdmEntityType nestedEntityType = navProperty.ToEntityType();
+
+            object obj;
+            if(parent.TryGetPropertyValue(navigationPropertyName, out obj))
```
```suggestion
            if (parent.TryGetPropertyValue(navigationPropertyName, out obj))
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+            return null;
+        }
+
+
```
```suggestion
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
+
+            if(originalList == null)
```
```suggestion
            if (originalList == null)
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultEdmPatchMethodHandler.cs:
```diff
@@ -0,0 +1,165 @@
+                    object value;
+                    if (item.TryGetPropertyValue(keyValue.Key, out value))
+                    {
+                        if (!Equals(value, keyValue.Value))
```
Wondering why we can't reverse the check here `if (Equals(value, keyValue.Value)` and return the item?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// This is the default ODataAPIHandler for CLR type. This calss has default Get, Create and Update
```
```suggestion
    /// This is the default ODataAPIHandler for CLR type. This class has default Get, Create and Update
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+    /// This is the default ODataAPIHandler for CLR type. This calss has default Get, Create and Update
+    /// and will do these actions. This will be used when the original collection to be Patched is provided.
+    /// </summary>
+    /// <typeparam name="TStructuralType"></typeparam>
```
Add a relevant comment to describe the `typeparam`

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+    /// and will do these actions. This will be used when the original collection to be Patched is provided.
+    /// </summary>
+    /// <typeparam name="TStructuralType"></typeparam>
+    internal class DefaultODataAPIHandler<TStructuralType> : ODataAPIHandler<TStructuralType> where TStructuralType :class
```
```suggestion
    internal class DefaultODataAPIHandler<TStructuralType> : ODataAPIHandler<TStructuralType> where TStructuralType : class
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+    /// <typeparam name="TStructuralType"></typeparam>
+    internal class DefaultODataAPIHandler<TStructuralType> : ODataAPIHandler<TStructuralType> where TStructuralType :class
+    {
+        Type _clrType;
+        ICollection<TStructuralType> originalList;
```
Consistency... How come one private field is declared with a leading `_` while the other is not?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+        public DefaultODataAPIHandler(ICollection<TStructuralType> originalList)
+        {
+            this._clrType = typeof(TStructuralType);
+            this.originalList = originalList?? new List<TStructuralType>();
```
```suggestion
            this.originalList = originalList ?? new List<TStructuralType>();
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+
+            try
+            {
+                if(originalList != null)
```
```suggestion
                if (originalList != null)
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+        public override ODataAPIResponseStatus TryGet(IDictionary<string, object> keyValues, out TStructuralType originalObject, out string errorMessage)
+        {
+            ODataAPIResponseStatus status = ODataAPIResponseStatus.Success;
+            errorMessage = string.Empty;
```
Why is the empty string preferred over `null` in the absence of an error?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+                    originalList = new List<TStructuralType>();
+                }
+
+                createdObject = Activator.CreateInstance(_clrType) as TStructuralType;
```
Are there any guarantees that the CLR type has a parameterless constructor?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+            try
+            {
+                TStructuralType originalObject = GetFilteredItem(keyValues);
+                originalList.Remove(originalObject);
```
How come unlike for `DefaultEdmODataAPIHandler` there's no check to verify that `originalObject` is non-null?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
+
+            if(originalList == null || originalList.Count == 0)
```
```suggestion
            if (originalList == null || originalList.Count == 0)
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DefaultODataAPIHandler.cs:
```diff
@@ -0,0 +1,157 @@
+
+                foreach (KeyValuePair<string, object> keyValue in keyValues)
+                {
+                    if (!Equals(propertyInfos[keyValue.Key].GetValue(item), keyValue.Value))
```
Wondering why we can't reverse the check here, i.e., `if (Equals(propertyInfos[keyValue.Key].GetValue(item), keyValue.Value))` and return the item?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="structuralType">The derived entity type or complex type for which the changes would be tracked.
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>         
+        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for Instance Annotations</param>
```
```suggestion
        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for instance annotations.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// </param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
+        /// properties. <c>null</c> means this entity type is not open.</param>        
+        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for Instance Annotations</param>
```
```suggestion
        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for instance annotations</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="structuralType">The derived entity type or complex type for which the changes would be tracked.
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
```
```suggestion
        /// <param name="dynamicDictionaryPropertyInfo">The property that is used as container dynamic
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>
+        /// <param name="updatableProperties"> Properties that can be updated</param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
```
```suggestion
        /// <param name="dynamicDictionaryPropertyInfo">The property that is used as container for dynamic
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="structuralType">The derived entity type or complex type for which the changes would be tracked.
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>
+        /// <param name="updatableProperties"> Properties that can be updated</param>
```
```suggestion
        /// <param name="updatableProperties">Properties that can be updated.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="updatableProperties"> Properties that can be updated</param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
+        /// properties. <c>null</c> means this entity type is not open.</param>
+        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for Instance Annotations</param>
```
```suggestion
        /// <param name="instanceAnnotationsPropertyInfo">The property that is used as container for instance annotations.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="structuralType">The derived entity type or complex type for which the changes would be tracked.
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>
+        /// <param name="updatableProperties"> Properties that can be updated</param>
```
```suggestion
        /// <param name="updatableProperties">Properties that can be updated.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <paramref name="structuralType"/> should be assignable to instances of <typeparamref name="TStructuralType"/>.
+        /// </param>
+        /// <param name="updatableProperties"> Properties that can be updated</param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
```
```suggestion
        /// <param name="dynamicDictionaryPropertyInfo">The property that is used as container for dynamic
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="updatableProperties"> Properties that can be updated</param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
+        /// properties. <c>null</c> means this entity type is not open.</param>
+        /// <param name="isComplexType">To determine if the entity is a complex type</param>
```
To keep the API a bit consistent, I'd make this the last argument instead of sandwiching it between `dynamicDictionaryPropertyInfo` and `instanceAnnotationsPropertyInfo`...

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
+        /// properties. <c>null</c> means this entity type is not open.</param>
+        /// <param name="isComplexType">To determine if the entity is a complex type</param>
+        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for Instance Annotations</param>
```
```suggestion
        /// <param name="instanceAnnotationsPropertyInfo">The property that is used as container for instance annotations.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaDeletedEntityObjectOfT.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <inheritdoc />
+        public IEdmNavigationSource NavigationSource { get; set; }  
+    }
+}
```
```suggestion
}

```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -79,6 +85,22 @@ public Delta(Type structuralType, IEnumerable<string> updatableProperties)
+        /// </param>
+        /// <param name="updatableProperties">The set of properties that can be updated or reset. Unknown property
+        /// names, including those of dynamic properties, are ignored.</param>
+        /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
```
```suggestion
        /// <param name="dynamicDictionaryPropertyInfo">The property that is used as container for dynamic
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -89,12 +111,13 @@ public Delta(Type structuralType, IEnumerable<string> updatableProperties)
         /// names, including those of dynamic properties, are ignored.</param>
         /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
         /// properties. <c>null</c> means this entity type is not open.</param>
+        /// <param name="isComplexType">Boolean value to determine if its a complex type</param>
```
We can't use `structuralType` properties or supported methods to check if it's a complex type instead of introducing this parameter?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -107,21 +130,29 @@ public Delta(Type structuralType, IEnumerable<string> updatableProperties)
         /// <param name="dynamicDictionaryPropertyInfo">The property info that is used as dictionary of dynamic
         /// properties. <c>null</c> means this entity type is not open.</param>
         /// <param name="isComplexType">Boolean value to determine if its a complex type</param>
+        /// <param name="instanceAnnotationsPropertyInfo">The property info that is used as container for Instance Annotations</param>
```
```suggestion
        /// <param name="instanceAnnotationsPropertyInfo">The property that is used as container for instance annotations.</param>
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -342,91 +434,53 @@ public void CopyChangedValues(TStructuralType original)
-
-                    // Recursively patch up the instance with the nested resources.
-                    deltaObject.CopyChangedValues(instance);
+                    IODataAPIHandler apiHandlerNested = apiHandler.GetNestedHandler(original, nestedResourceName);
```
```suggestion
                    IODataAPIHandler apiHandlerNested = apiHandler.GetNestedHandler(original, nestedResourceName);
```
```suggestion
                    IODataAPIHandler nestedApiHandler = apiHandler.GetNestedHandler(original, nestedResourceName);
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -465,6 +519,51 @@ public void Patch(TStructuralType original)
+        {
+            IODataAPIHandler refapiHandler = apiHandlerFactory.GetHandler(ODataIdContainer.ODataIdNavigationPath);
+
+            if (refapiHandler != null)
```
```suggestion
            if (refApiHandler != null)
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -465,6 +519,51 @@ public void Patch(TStructuralType original)
+
+            if (refapiHandler != null)
+            {
+                ODataAPIHandler<TStructuralType> refapiHandlerOfT = refapiHandler as ODataAPIHandler<TStructuralType>;
```
```suggestion
                ODataAPIHandler<TStructuralType> refApiHandlerOfT = refapiHandler as ODataAPIHandler<TStructuralType>;
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -465,6 +519,51 @@ public void Patch(TStructuralType original)
+                string error;
+
+                //Checking to get the referenced entity, get the properties and apply it on original object
+                if (refapiHandlerOfT.TryGet(ODataIdContainer.ODataIdNavigationPath.GetNavigationPathItems().Last().KeyProperties, out referencedObj, out error) == ODataAPIResponseStatus.Success)
```
Is `ODataIdContainer.ODataIdNavigationPath.GetNavigationPathItems().Last()` guaranteed to return a non-null?

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs:
```diff
@@ -793,4 +949,4 @@ private bool TrySetNestedResourceInternal(string name, object deltaNestedResourc
         }
     }
-}
+}
```
```suggestion
}

```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+    [NonValidatingParameterBinding]
+    public class DeltaSet<TStructuralType> : Collection<IDeltaSetItem>, IDeltaSet where TStructuralType : class
+    {        
+        private Type _clrType;
+        IList<string> _keys;
+        NavigationPath _navigationPath;
```
Consistency: One field has the `private` access modifier while others do not...

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+            base.InsertItem(index, item);
+        }
+
+
```
```suggestion
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+            _navigationPath = new NavigationPath(_clrType.Name, null);
+        }
+
+    
```
```suggestion
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+            return CopyChangedValues(apiHandler);
+        }
+
+
```
```suggestion
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+            return CopyChangedValues(apiHandlerOfT, apiHandlerFactory);
+        }
+
+
```
```suggestion
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+
+                        if (apiHandlerOfT.TryDelete(keyValues, out errorMessage) != ODataAPIResponseStatus.Success)
+                        {
+                            //Handle Failed Operation - Delete                           
```
TODO

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+        {
+            Type type = typeof(Delta<>).MakeGenericType(_clrType);
+
+            Delta<TStructuralType> deltaObject = Activator.CreateInstance(type, _clrType, null, null,false,
```
```suggestion
            Delta<TStructuralType> deltaObject = Activator.CreateInstance(type, _clrType, null, null, false,
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+        //This is for ODL to work to set id as empty, because if there are missing keys, id wouldnt be set and we need to set it as empty.
+        private static void ValidateForDeletedEntityId(IList<string> keys, DeltaDeletedEntityObject<TStructuralType> edmDeletedObject)
+        {
+            bool hasnullKeys = false;
```
```suggestion
            bool hasNullKeys = false;
```

gathogojr commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,346 @@
+            }
+        }
+    }
+}
```
```suggestion
}

```

Sreejithpin committed:
Updates

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+        /// Initializes a new instance of the <see cref="NavigationPath"/> class.
+        /// </summary>        
+        /// <param name="pathSegments">Pathsegment collection</param>
+        public NavigationPath(ReadOnlyCollection<ODataPathSegment> pathSegments)
```
>ReadOnlyCollection [](http://example.com/codeflow?start=30&length=18)

Does it have to be a ReadOnlyCollection?

Is this the right public abstraction?  Will they typically have a readonlycollection of odatapathsegments, or will they have a Uri?

Sreejithpin replied:
ODataPath.Segments is actually readonly collection (defined)

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+        /// <param name="pathSegments">Pathsegment collection</param>
+        public NavigationPath(ReadOnlyCollection<ODataPathSegment> pathSegments)
+        {
+             Debug.Assert(pathSegments.Count > 0);
```
>Debug.Assert(pathSegments.Count > 0); [](http://example.com/codeflow?start=13&length=37)

if this is a public API, then we should validate (not assert) that pathSegments is non null and has count >0.

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+             Debug.Assert(pathSegments.Count > 0);
+
+            _pathSegments = pathSegments;
+            _navigationPathName = pathSegments.First().Identifier;
```
>_navigationPathName [](http://example.com/codeflow?start=12&length=19)

Is this always right? 
What does the collection of path segments represent? Will it always start with a top level entityset or singleton (or function import?) where/how is _navigationPathName used?

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/ODataIdContainer.cs:
```diff
@@ -0,0 +1,37 @@
+        NavigationPath ODataIdNavigationPath { set; get; }
+
+        /// <summary>
+        /// OdataId path string
```
>OdataId path string [](http://example.com/codeflow?start=12&length=19)

is this always a path string, or could it be opaque?

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+        /// <summary>
+        /// Gets the NavigationPath name
+        /// </summary>
+        public string NavigationPathName { get { return _navigationPathName; } }
```
is this still used anywhere?

Sreejithpin replied:
Will be used in the GetHandler of handler factory which willbe used in the case of @odata.id

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+        /// To Get ODataId in Parsed format
+        /// </summary>
+        /// <returns>Array of PathItems</returns>
+        public PathItem[] GetNavigationPathItems()
```
>GetNavigationPathItems [](http://example.com/codeflow?start=26&length=22)

This is the main logic/use of the class, right? getting the path items?

Maybe make this just public PathItem[] PathItems?

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/NavigationPath.cs:
```diff
@@ -0,0 +1,90 @@
+        /// Initializes a new instance of the <see cref="NavigationPath"/> class.
+        /// </summary>        
+        /// <param name="pathSegments">Pathsegment collection</param>
+        public NavigationPath(ReadOnlyCollection<ODataPathSegment> pathSegments)
```
>NavigationPath [](http://example.com/codeflow?start=15&length=14)

Do we need a public constructor at all for this (at least initially), or will we always create the NavigationPath object?

Sreejithpin replied:
Did you mean internal constructor? Initially internal will be fine to see if any user wants to use  /create it

mikepizzo replied:
Yes; if we can make internal that would be great.

Sreejithpin replied:
Ok doing it

mikepizzo committed:
Draft PR to experiment with refactoring
-remove NavigationPath from ODataIdContainer
-make it easy for developer to create a NavigationPath from id/Model
-Add model to APIHandlerFactory base class as a way to provide the model to the methods that require it
-Experiment with setting the id during deserialization in cases where only key values are provided. Should we get this from ODL?
-use the actual path for nested resources in readContext

todo:
-rationalize routing odatapath and uriparser odatapath usage
-build out logic/add tests for the path in nested resources for readContext
-should we get the id for key properties during deserialization from ODL?
-handle edge cases throughout

mikepizzo committed:
Improve logic for setting id on resourceWrapper using path and key values, if specified

mikepizzo commented on src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs:
```diff
@@ -358,6 +419,243 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+        /// <param name="resourceWrapper">The resource wrapper.</param>
+        /// <param name="readContext">The read context.</param>
+        /// <returns>The resource wrapper.</returns>
+        private ODataResourceWrapper UpdateResourceWrapper(ODataResourceWrapper resourceWrapper, ODataDeserializerContext readContext)
```
>UpdateResourceWrapper [](http://example.com/codeflow?start=37&length=21)

This method name and description are vague; they don't tell me how the resource wrapper is updated.  Maybe rename "SetKeyProperties"?

Sreejithpin committed:
Update ODataResourceDeserializer.cs

Sreejithpin committed:
New PR for Serializers for Bulk Operation with Updated ContextPath (#2618)

* Temp commit

* Updates

Sreejithpin committed:
Updates

Sreejithpin committed:
updates for casttype

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,344 @@
+        /// <param name="keys">List of key names for the type</param>        
+        public DeltaSet(IList<string> keys)           
+        {
+            _keys = keys;            
```
It seems that later on we assume that `_keys` is not `null`; since this is a public API, it's probably best to throw if `keys` is `null`.

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,344 @@
+
+        private DeltaSet<TStructuralType> CreateDeltaSet()
+        {
+            Type type = typeof(DeltaSet<>).MakeGenericType(_clrType);
```
Is it not legal to just write `typeof(DeltaSet<TStructuralType>)`?

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs:
```diff
@@ -0,0 +1,344 @@
+
+            if (hasnullKeys)
+            {               
+                edmDeletedObject.Id = new Uri(string.Empty);                
```
Small optimization, I think that `Uri` is immutable (or functionally immutable), so we can reduce object allocations if we have a singleton definition of `new Uri(string.Empty)`

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollection.cs:
```diff
@@ -60,6 +70,321 @@ private void Initialize(IEdmEntityType entityType)
+        internal EdmChangedObjectCollection CopyChangedValues(EdmODataAPIHandler apiHandler, ODataEdmAPIHandlerFactory apiHandlerFactory = null)
+        {
+            EdmChangedObjectCollection changedObjectCollection = new EdmChangedObjectCollection(_entityType);
+            IEdmStructuralProperty[] keys = _entityType.Key().ToArray();
```
I think there must be a better approach than this. I was initially suspicious because of the `ToArray` call, but then I saw it makes sense because it is possible for `keys` to be enumerated multiple times. However, at least for the failed case, we are doing the same `.Name` deference for the keys. Maybe here we should do `_entityType.Key().Select(key => key.Name).ToArray()` so that we don't have to deference the name so many times?

Sreejithpin replied:
Good one

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/EdmODataAPIHandler.cs:
```diff
@@ -0,0 +1,46 @@
+        /// <param name="createdObject">The created object (Typeless)</param>
+        /// <param name="errorMessage">Any error message in case of an exception</param>
+        /// <returns>The status of the TryCreate Method, statuses are <see cref="ODataAPIResponseStatus"/></returns>
+        public abstract ODataAPIResponseStatus TryCreate(IEdmChangedObject changedObject, out IEdmStructuredObject createdObject, out string errorMessage);
```
I would prefer the return type be a structure that has both `ODataAPIResponseStatus` and `errorMessage` as properties. I think it makes the pattern of having both return values clearer. Do you have any thoughts on this?

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/EdmODataAPIHandler.cs:
```diff
@@ -0,0 +1,46 @@
+    /// Handler Class to handle users methods for create, delete and update.
+    /// This is the handler for data modification where there is no CLR type.
+    /// </summary>
+    internal abstract class EdmODataAPIHandler
```
Why can't this be an `interface` instead of `abstract class`?

Sreejithpin replied:
To maintain pattern similar to internal abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where TStructuralType : class

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/EdmStructuredObject.cs:
```diff
@@ -224,8 +225,15 @@ internal static object GetDefaultValue(IEdmTypeReference propertyType)
                 {
+
+                    bool hasDefaultConstructor = (!clrType.IsClass) || clrType.GetConstructors(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static).
+                                                 Any(x => x.GetParameters().Count() == 0);
```
`GetParameters` returns an array, so calling `Length` instead of `Count()` should save us a cast

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs:
```diff
@@ -348,14 +414,361 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+
+                if (keySegment != null)
+                {
+                    foreach (KeyValuePair<string, object> key in keySegment.Keys)
```
I don't really see the value of doing this over `keySegment.Keys.Select(key => new ODataProperty { Name = key.Key, Value = key.Value }).ToList()`. If `KeySegment` has a count on it, then maybe it makes sense to initialize a new list with that capacity, but otherwise this just seems verbose.

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs:
```diff
@@ -348,14 +414,361 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+                    if (property == null && !readContext.DisableCaseInsensitiveRequestPropertyBinding)
+                    {
+                        //try case insensitive
+                        List<ODataProperty> candidates = resourceWrapper.ResourceBase.Properties.Where(p => String.Equals(p.Name, keyName, StringComparison.InvariantCultureIgnoreCase)).ToList();
```
I strongly suggest implementing an extension method for this. I always thought it was weird that this wasn't the functionality for `SingleOrDefault` in the first place.

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataDeltaFeedSerializer.cs:
```diff
@@ -456,14 +510,22 @@ public virtual void WriteDeltaDeletedEntry(object graph, ODataWriter writer, ODa
+
+                if (deletedResource != null)
+                {
+                    await writer.WriteStartAsync(deletedResource);
```
Do we need to continue execution on the same thread that awaited? I wonder if we want to call `ConfigureAwait(false)` here

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/IDeltaDeletedEntityObject.cs:
```diff
@@ -0,0 +1,35 @@
+        /// <summary>
+        /// Optional. Either deleted, if the entity was deleted (destroyed), or changed if the entity was removed from membership in the result (i.e., due to a data change).
+        /// </summary>
+        DeltaDeletedEntryReason? Reason { get; set; }
```
Should this be "entity" instead of "entry"?

corranrogue9 commented on src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs:
```diff
@@ -0,0 +1,54 @@
+    /// Handler Class to handle users methods for create, delete and update.
+    /// This is the handler for data modification where there is a CLR type.
+    /// </summary>
+    internal abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where TStructuralType : class
```
It seems weird that we are constraining the generic type parameter to just `class`. Why is that restriction necessary?

Sreejithpin replied:
It is used to handle delta request and it should be only for class types

corranrogue9 commented on test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Build.AspNet/Formatter/ODataEntityReferenceLinkE2ETests.cs:
```diff
@@ -0,0 +1,276 @@
+            "\"Author@odata.bind\":\"Authors(1)\"}";
+
+            string Uri = BaseAddress + "/odata/Books";
+            HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Post, Uri);
```
This implements `IDisposable`. Please wrap it in a `using` block

corranrogue9 commented on test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Build.AspNet/Formatter/ODataEntityReferenceLinkE2ETests.cs:
```diff
@@ -0,0 +1,276 @@
+            request.Content.Headers.ContentType = MediaTypeWithQualityHeaderValue.Parse("application/json");
+
+            // Act
+            HttpResponseMessage response = await Client.SendAsync(request);
```
This also implements `IDisposable`. Please check for other types in the tests as well

Sreejithpin committed:
updates
